text/0058-valuecastable-format.md

```diff
@@ -0,0 +1,93 @@
+
+# Core support for `ValueCastable` formatting
+
+## Summary
+[summary]: #summary
+
+`Format` hooks are added, allowing custom formatting to be implemented for `ValueCastable`s.
+
+This RFC is only about adding hook support for custom `ShapeCastable`s. Providing actual formatting implementation for `lib.data` and `lib.enum` is left for a future RFC.
+
+## Motivation
+[motivation]: #motivation
+
+Custom types, like enums and data structures, will not make immediate sense to the user when printed as raw bit patterns. However, it is often desirable to print them out with the `Print` statement. Thus, an extension point for custom formatting needs to be added to the `Format` machinery.
+
+## Guide-level explanation
+[guide-level-explanation]: #guide-level-explanation
+
+`ShapeCastable` subtypes can define their own ways of formatting values. For example, consider the following:
+
+```py
+class Opcode(amaranth.lib.enum.Enum, shape=unsigned(2)):
+    ADD = 0
+    SUB = 1
+    XOR = 2
+
+op = Signal(Opcode)
+m.d.comb += [
+    op.eq(Opcode.XOR)
+    Print(Format("current opcode: {}", op)),
+]
+```
+
+It is desirable that this example would print something like `"XOR"`, instead of `"2"`. Such functionality can be added on `ShapeCastable` subclasses by overriding the `format` method.
+
+However, sometimes it is also useful to print the raw value of such a signal. A shorthand is provided for that purpose:
+
+```py
+m.d.comb += Print(Format("current opcode: {op} (raw: {op!v})", op=op))
+```
+
+## Reference-level explanation
+[reference-level-explanation]: #reference-level-explanation
+
+A new overridable method is added to `ShapeCastable`:
+
+- `format(self, value: ValueCastable, format_desc: str) -> Format`
+
+When a `ValueCastable` is formatted without specifying a conversion (ie. `"!r"` or `"!v"`):
+
+- `shape()` is called
+- if the shape is a plain `Shape`, `Value.cast()` is called on the value-castable, and formatting proceeds as if it was a plain value
+- otherwise (the shape is a `ShapeCastable`), the value-castable being formatted and the format descriptor (the part after `:` in `{}`, if any) are passed directly to `shape.format()`, and the result (which must be a `Format` object) is included directly in place of the format specifier
+
+The default implementation of `format` raises a `TypeError` exception.
+
+A new conversion, `!v`, is added to `Format` (with syntax analogous to `!r`). When specified, the value being formatted is converted through `Value.cast` before proceeding with further formatting. It can be used to print the raw value of value-castables. It is a no-op when used with a plain `Value`.
+
+An implementation of `__format__` is added to the `ValueCastable` base class that always raises a `TypeError`, directing the user to `Format` instead (like the one that already exists on plain `Value`).
+
+## Drawbacks
+[drawbacks]: #drawbacks
+
+None.
```

Comment:
The fact that a new name `format` is reserved on `ShapeCastable` and a default implementation for `__format__` is added to `ValueCastable` are definitely drawbacks as we are trying to keep those interfaces slim.

I feel especially conflicted about `__format__` on `ValueCastable`. That will be the first method with an actual implementation added on that interface. Is this the right thing to do?